sql/src/main/java/io/crate/operation/reference/sys/node/DiscoveryNodeContextFieldResolver.java

```diff
@@ -0,0 +1,251 @@
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.sys.node;
+
+import com.google.common.collect.ImmutableMap;
+import io.crate.Build;
+import io.crate.Version;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.ReferenceIdent;
+import io.crate.metadata.sys.SysNodesTableInfo;
+import io.crate.monitor.ExtendedNodeInfo;
+import io.crate.monitor.ThreadPools;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.lucene.BytesRefs;
+import org.elasticsearch.common.transport.InetSocketTransportAddress;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.util.Consumer;
+import org.elasticsearch.common.util.concurrent.XRejectedExecutionHandler;
+import org.elasticsearch.monitor.jvm.JvmService;
+import org.elasticsearch.monitor.os.OsService;
+import org.elasticsearch.node.service.NodeService;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.*;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+
+@Singleton
+public class DiscoveryNodeContextFieldResolver {
+
+    private final OsService osService;
+    private final JvmService jvmService;
+    private final ClusterService clusterService;
+    private final NodeService nodeService;
+    private final ThreadPool threadPool;
+    private final ExtendedNodeInfo extendedNodeInfo;
+
+    @Inject
+    public DiscoveryNodeContextFieldResolver(ClusterService clusterService,
+                                             OsService osService,
+                                             NodeService nodeService,
+                                             JvmService jvmService,
+                                             ThreadPool threadPool,
+                                             ExtendedNodeInfo extendedNodeInfo) {
+        this.osService = osService;
+        this.jvmService = jvmService;
+        this.clusterService = clusterService;
+        this.nodeService = nodeService;
+        this.threadPool = threadPool;
+        this.extendedNodeInfo = extendedNodeInfo;
+    }
+
+    public DiscoveryNodeContext resolveForColumnIdents(Collection<ReferenceIdent> referenceIdents) {
+        DiscoveryNodeContext context = DiscoveryNodeContext.newInstance();
+        if (referenceIdents == null || referenceIdents.isEmpty()) {
```

Comment:
is the argument nullable? If so mark it as such. Usually we don't expect collections or lists to be nullable but only empty